File: app/src/main/java/com/galtashma/parsedashboard/screens/SingleAppParseActivity.java
package com.galtashma.parsedashboard.screens;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.ListView;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.galtashma.parsedashboard.Const;
import com.galtashma.parsedashboard.R;
import com.galtashma.parsedashboard.adapters.ParseClassesAdapter;
import com.parse.ParseSchema;
import com.vlonjatg.progressactivity.ProgressRelativeLayout;

import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import bolts.Continuation;


public class SingleAppParseActivity extends AppCompatActivity {

    private ParseClassesAdapter adapter;
    private ProgressRelativeLayout statefulLayout;

    private Map<String, ParseSchema> schemas = new HashMap<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_single_app);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        String appName = "";

        Bundle extra = getIntent().getExtras();
        if (extra == null) {
            extra = savedInstanceState;
        }

        if (extra != null && extra.containsKey(Const.BUNDLE_KEY_PARSE_APP_NAME)) {
            appName = extra.getString(Const.BUNDLE_KEY_PARSE_APP_NAME);
        }

        setTitle(appName);

        statefulLayout = findViewById(R.id.stateful_layout);
        adapter = new ParseClassesAdapter(this);
        adapter.setListener(schema -> showTable(schema.getName()));
        ListView listView = findViewById(R.id.list_view);
        listView.setAdapter(adapter);
        fetchSchemasAsync();
    }

    private void fetchSchemasAsync() {
        statefulLayout.showLoading();
        ParseSchema.getParseSchemasAsync().continueWith((Continuation<List<ParseSchema>, Void>) task -> {
            if (task.isFaulted() || task.isCancelled()) {
                showErrorOnUIThread(getString(R.string.schemas_screen_error_title), task.getError());
                Log.e("ParseDashboard","Error fetching from "  + " parse server.", task.getError());
                return null;
            }

            Log.i(Const.TAG, "found schemas " + task.getResult());
            List <ParseSchema> s = task.getResult();
            updateListOnUIThread(s);
            for (ParseSchema ps : s) {
                schemas.put(ps.getName(), ps);
            }

            return null;
        });
    }

    private void updateListOnUIThread(final List<ParseSchema> schemasList) {
        runOnUiThread(() -> {
            if (schemasList.size() == 0) {
                showEmptyState();
                return;
            }
            statefulLayout.showContent();
            adapter.clear();
            adapter.addAll(schemasList);
            adapter.notifyDataSetChanged();
        });
    }

    private void showErrorOnUIThread(final String title, final Exception e) {
        runOnUiThread(() -> statefulLayout.showError(R.drawable.ic_parse_24dp, title, e.getLocalizedMessage(), "Retry", view -> fetchSchemasAsync()));
    }

    private void showEmptyState() {
        statefulLayout.showEmpty(R.drawable.ic_parse_24dp, getString(R.string.empty_state_schemas_screen_short), getString(R.string.empty_state_schemas_screen_long));
    }

    private void showTable(String tableName) {
        Intent intent = new Intent(this, SingleClassParseActivity.class);
        intent.putExtra(Const.BUNDLE_KEY_CLASS_NAME, tableName);
        if (schemas.containsKey(tableName)) {
            int size = schemas.get(tableName).getFields().size();
            String[] fieldsArr = new String[size];

            Iterator<String> it = schemas.get(tableName).getFields().keySet().iterator();
            for (int i=0; i<size; i++) {
                if (it.hasNext()) {
                    fieldsArr[i] = it.next();
                }
            }
            intent.putExtra(Const.BUNDLE_KEY_CLASS_FIELDS_NAME, fieldsArr);
        }
        this.startActivityForResult(intent, 1);
    }
}

File: app/src/main/java/com/galtashma/parsedashboard/screens/SingleAppParseActivity.kt
package com.galtashma.parsedashboard.screens

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.ListView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import com.galtashma.parsedashboard.Const
import com.galtashma.parsedashboard.R
import com.galtashma.parsedashboard.adapters.ParseClassesAdapter
import com.parse.ParseSchema
import com.vlonjatg.progressactivity.ProgressRelativeLayout

/**
 * Created by gal on 3/16/18, rewritten by Cyb3rKo on 05/12/22.
 */

class SingleAppParseActivity : AppCompatActivity() {
    private lateinit var adapter: ParseClassesAdapter
    private lateinit var statefulLayout: ProgressRelativeLayout

    private val schemas = hashMapOf<String, ParseSchema>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_single_app)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)

        var extra = intent.extras
        if (extra == null) {
            extra = savedInstanceState
        }

        val appName = if (extra != null && extra.containsKey(Const.BUNDLE_KEY_PARSE_APP_NAME)) {
            extra.getString(Const.BUNDLE_KEY_PARSE_APP_NAME)
        } else ""

        title = appName

        statefulLayout = findViewById(R.id.stateful_layout)
        adapter = ParseClassesAdapter(this)
        adapter.setListener(object: ParseClassesAdapter.OnClickListener {
            override fun onSchemaCLicked(schema: ParseSchema) {
                showTable(schema.name)
            }
        })
        val listView = findViewById<ListView>(R.id.list_view)
        listView.adapter = adapter
        fetchSchemasAsync()
    }

    private fun fetchSchemasAsync() {
        statefulLayout.showLoading()
        ParseSchema.getParseSchemasAsync().continueWith { task ->
            if (task.isFaulted || task.isCancelled) {
                showErrorOnUIThread(getString(R.string.schemas_screen_error_title), task.error)
                Log.e("ParseDashboard","Error fetching from parse server.", task.error)
                return@continueWith null
            }

            Log.i(Const.TAG, "Found schemas: ${task.result}")
            val s = task.result
            updateListOnUIThread(s)
            s.forEach { ps ->
                schemas[ps.name] = ps
            }

            return@continueWith null
        }
    }

    private fun updateListOnUIThread(schemasList: List<ParseSchema>) {
        runOnUiThread {
            if (schemasList.isEmpty()) {
                showEmptyState()
                return@runOnUiThread
            }
            statefulLayout.showContent()
            adapter.clear()
            adapter.addAll(schemasList)
            adapter.notifyDataSetChanged()
        }
    }

    private fun showErrorOnUIThread(title: String, e: Exception) {
        runOnUiThread {
            statefulLayout.showError(
                R.drawable.ic_parse_24dp,
                title,
                e.localizedMessage,
                "Retry"
            ) {
                fetchSchemasAsync()
            }
        }
    }

    private fun showEmptyState() {
        statefulLayout.showEmpty(
            R.drawable.ic_parse_24dp,
            getString(R.string.empty_state_schemas_screen_short),
            getString(R.string.empty_state_schemas_screen_long)
        )
    }

    private fun showTable(tableName: String) {
        val intent = Intent(this, SingleClassParseActivity::class.java)
        intent.putExtra(Const.BUNDLE_KEY_CLASS_NAME, tableName)
        if (schemas.containsKey(tableName)) {
            val fields = schemas[tableName]?.fields?.keys?.toTypedArray()
            if (fields != null) {
                intent.putExtra(Const.BUNDLE_KEY_CLASS_FIELDS_NAME, fields)
                this.startActivityForResult(intent, 1)
            }
        }
    }
}
